File: matlab/src/api/moment/moment_joint_raw_from_central.m
function m = moment_joint_raw_from_central(mc, mu)
% m = moment_joint_raw_from_central(mc, mu)
%
% Converts the joint central moments of a random vector into the joint power
% (raw) moments, by the multi-index binomial theorem,
%
%   m_(i) = sum_(k<=i) prod_j nchoosek(i_j,k_j) mu_j^(i_j-k_j) * mc_(k)
%
% The mean vector must be supplied separately, since the first-order central
% moments are zero and carry no information on it. Inverse of
% moment_joint_central_from_raw.
%
% Input:
%   mc: array of size (n_1+1)x...x(n_d+1) holding the joint central moments
%   mu: vector of length d holding the means E[N_1],...,E[N_d]
%
% Output:
%   m: array of the same size as mc holding the joint power moments
%
% Example:
%   m = moment_joint_raw_from_central(moment_joint_central_from_raw(m0), mu);
%
% Reference:
% A. Heindl and A. van de Liefvoort. Moment conversions for discrete
% distributions. PMCCS, 2003, Section 4.

sz = moment_tensorsize(mc);
d = numel(sz);
if numel(mu) ~= d
    line_error(mfilename,'The mean vector mu must have one entry per dimension of mc.');
end
m = mc;
for mode = 1:d
    n = sz(mode)-1;
    T = zeros(n+1,n+1);
    for i = 0:n
        for k = 0:i
            T(i+1,k+1) = nchoosek(i,k) * mu(mode)^(i-k);
        end
    end
    m = moment_tensortrans(m, T, mode);
end
end

File: jar/src/main/java/jline/api/moment/Moment_joint_raw_from_central.java
package jline.api.moment;

import jline.util.matrix.Matrix;

/**
 * Joint power (raw) moments from joint central moments.
 *
 * <p>Inverts {@link Moment_joint_central_from_raw} by the multi-index binomial
 * theorem,
 *
 * <pre>
 *   m_(i) = sum_(k&lt;=i) prod_j nchoosek(i_j,k_j) mu_j^(i_j-k_j) * mc_(k)
 * </pre>
 *
 * <p>The mean vector must be supplied separately, since the first-order central
 * moments are zero and carry no information on it.
 *
 * <p>Reference:
 * A. Heindl and A. van de Liefvoort. Moment conversions for discrete
 * distributions. PMCCS, 2003, Section 4.
 *
 * @since LINE 3.0
 */
public final class Moment_joint_raw_from_central {
    private Moment_joint_raw_from_central() {}

    /**
     * Converts joint central moments into joint power moments.
     *
     * @param mc flattened joint central moments in row-major order
     * @param dims extents of the array, dims[j] = n_j+1
     * @param mu means E[N_1],...,E[N_d], one per dimension
     * @return flattened joint power moments, same layout
     */
    public static double[] moment_joint_raw_from_central(double[] mc, int[] dims, double[] mu) {
        if (mu.length != dims.length) {
            throw new IllegalArgumentException("moment_joint_raw_from_central: The mean vector mu "
                    + "must have one entry per dimension of mc.");
        }
        double[] out = mc;
        for (int mode = 0; mode < dims.length; mode++) {
            int n = dims[mode] - 1;
            Matrix T = new Matrix(n + 1, n + 1);
            for (int i = 0; i <= n; i++) {
                for (int k = 0; k <= i; k++) {
                    T.set(i, k, Moment_binotrans.nchoosek(i, k) * Math.pow(mu[mode], i - k));
                }
            }
            out = Moment_tensortrans.moment_tensortrans(out, dims, T, mode);
        }
        return out;
    }
}
